src/compiler/checker.ts

```diff
@@ -14624,68 +14653,49 @@ namespace ts {
         }
 
-        function getReferenceParent(source: Node) {
-            if (source.kind === SyntaxKind.PropertyAccessExpression) {
-                return (<PropertyAccessExpression>source).expression;
-            }
-            if (source.kind === SyntaxKind.BindingElement) {
-                const ancestor = source.parent.parent;
-                return ancestor.kind === SyntaxKind.VariableDeclaration ? (<VariableDeclaration>ancestor).initializer : ancestor;
-            }
-            return undefined;
-        }
-
         function containsMatchingReference(source: Node, target: Node) {
-            let parent = getReferenceParent(source);
-            while (parent) {
-                if (isMatchingReference(parent, target)) {
+            while (isAccessExpression(source)) {
+                source = source.expression;
+                if (isMatchingReference(source, target)) {
                     return true;
                 }
-                parent = getReferenceParent(parent);
             }
+            return false;
         }
 
         // Return true if target is a property access xxx.yyy, source is a property access xxx.zzz, the declared
         // type of xxx is a union type, and yyy is a property that is possibly a discriminant. We consider a property
         // a possible discriminant if its type differs in the constituents of containing union type, and if every
         // choice is a unit type or a union of unit types.
         function containsMatchingReferenceDiscriminant(source: Node, target: Node) {
-            return target.kind === SyntaxKind.PropertyAccessExpression &&
-                containsMatchingReference(source, (<PropertyAccessExpression>target).expression) &&
-                isDiscriminantProperty(getDeclaredTypeOfReference((<PropertyAccessExpression>target).expression), (<PropertyAccessExpression>target).name.escapedText);
+            return isAccessExpression(target) &&
+                containsMatchingReference(source, target.expression) &&
+                isDiscriminantProperty(getDeclaredTypeOfReference(target.expression), getAccessedPropertyName(target));
         }
 
         function getDeclaredTypeOfReference(expr: Node): Type | undefined {
             if (expr.kind === SyntaxKind.Identifier) {
                 return getTypeOfSymbol(getResolvedSymbol(<Identifier>expr));
             }
-            if (expr.kind === SyntaxKind.PropertyAccessExpression) {
-                const type = getDeclaredTypeOfReference((<PropertyAccessExpression>expr).expression);
-                return type && getTypeOfPropertyOfType(type, (<PropertyAccessExpression>expr).name.escapedText);
+            if (isAccessExpression(expr)) {
+                const type = getDeclaredTypeOfReference(expr.expression);
+                if (type) {
+                    const propName = getAccessedPropertyName(expr);
+                    return propName && getTypeOfPropertyOfType(type, propName);

```

Comment:
`""` is falsy, and `getAccessedPropertyName` returns a stringish thing. This needs to use `typeof`.

Reply:
Ya, that's a problem.